sql/lock_subsystem_test.go

```diff
@@ -0,0 +1,174 @@
+package sql
+
+import (
+	"github.com/stretchr/testify/assert"
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+)
+
+const testLockName = "my_lock"
+
+type lockDiffs struct {
+	missing []string
+	extra []string
+}
+
+func getLockDiffs(ctx *Context, testLockNames ...string) *lockDiffs {
+	var missing []string
+	var extra [] string
+
+	locks := make(map[string]bool)
+	for _, name := range testLockNames {
+		locks[name] = true
+	}
+
+	userLocks := make(map[string]bool)
+	_ = ctx.IterLocks(func(name string) error {
+		userLocks[name] = true
+
+		if !locks[name] {
+			extra = append(extra, name)
+		}
+		return nil
+	})
+
+	for name := range locks {
+		if !userLocks[name] {
+			missing = append(missing, name)
+		}
+	}
+
+	if len(missing) == 0 && len(extra) == 0 {
+		return nil
+	}
+
+	return &lockDiffs{missing: missing, extra: extra}
+}
+
+func TestLock(t *testing.T) {
+	ls := NewLockSubsystem()
+	user1Ctx := NewEmptyContext()
+	user2Ctx := NewEmptyContext()
+
+	err := ls.Lock(user1Ctx, testLockName, 0)
+	assert.NoError(t, err)
+	assert.Nil(t, getLockDiffs(user1Ctx, testLockName))
+
+	err = ls.Lock(user1Ctx, testLockName, 0)
+	assert.NoError(t, err)
+	assert.Nil(t, getLockDiffs(user1Ctx, testLockName))
+
+	err = ls.Lock(user2Ctx, testLockName, 0)
+	assert.Error(t, err)
+	assert.Nil(t, getLockDiffs(user2Ctx))
+
+	err = ls.Unlock(user1Ctx, testLockName)
+	assert.NoError(t, err)
+	assert.Nil(t, getLockDiffs(user1Ctx, testLockName))
+
+	err = ls.Lock(user2Ctx, testLockName, 0)
+	assert.Error(t, err)
+	assert.Nil(t, getLockDiffs(user2Ctx))
+
+	err = ls.Unlock(user1Ctx, testLockName)
+	assert.NoError(t, err)
+	assert.Nil(t, getLockDiffs(user1Ctx))
+
+	err = ls.Lock(user1Ctx, testLockName, 0)
+	assert.NoError(t, err)
+	assert.Nil(t, getLockDiffs(user1Ctx, testLockName))
+
+	err = ls.Lock(user2Ctx, testLockName, 0)
+	assert.Error(t, err)
+	assert.Nil(t, getLockDiffs(user2Ctx))
+}
+
+func TestRace(t *testing.T) {
```

Comment:
This is fine, but there's specific go tooling for finding races that you should figure out how to effectively tap into with a test.